### PR TITLE
Add a flag for enabling config watch

### PR DESCRIPTION
**Pull request by swiatekm-sumo**
**Description:**
Add a flag to control configuration watching. 

Technically this is a breaking change, as the flag defaults to false, and this is currently unconditionally enabled. In practice, none of the currently available config sources implement notifications, so users shouldn't see any difference unless they're using a distribution which specifically added such a config source. This being disabled by default will also allow changes like #5945 to be merged without affecting users unless they opt-in to the behaviour.

I'm open to suggestions for the flag name.

**Link to tracking Issue:** #5966 

**Testing:**
Added a unit test for the default (disabled), and updated the one checking for the other case.

**Documentation:** 
Do I need to update this in https://github.com/open-telemetry/opentelemetry.io?

### CODE REVIEW

File: service/flags.go
Comment (bogdandrutu) [inline service/flags.go:27]: "config-autoreload"?

I don't like necessarily the "hot" part of the name. "auto" sounds better to me.

Also important to mention what is reloaded :)

File: service/collector.go
Comment (bogdandrutu) [inline service/collector.go:178]: Not sure about this, it may spam the logging system.

Comment (swiatekm-sumo) [reply service/collector.go:178]: It might, but I think it's still useful information that config sources keep changing and the collector is ignoring them. In most cases I think this will alert users that they need to enable reloading if they want it.

File: service/flags.go
Comment (swiatekm-sumo) [reply service/flags.go:27]: Hm, how about just `--watch-config`? That's consistent with our internal terminology, and reasonably descriptive. I'm fine with `--enable-auto-reload` as well though.

File: service/flags.go
Comment (swiatekm-sumo) [reply service/flags.go]: Done. Would be less awkward if `flags` exposed its own `boolValue`, but it's not too much additional code.

File: service/flags.go
Comment (bogdandrutu) [inline service/flags.go:89]: flagSet.BoolVar(
		new(bool),
		watchConfigFlag,
		"Enable config watch for configuration sources.",
	)

File: service/flags.go
Comment (bogdandrutu) [inline service/flags.go:104]: return flagSet.Lookup(watchConfigFlag).Value.(flag.Getter).Get().(bool)

File: service/collector.go
Comment (bogdandrutu) [reply service/collector.go]: And if they don't want auto-reload? Do we just spam with "warnings"?

File: service/collector.go
Comment (bogdandrutu) [inline service/collector.go:175]: I thought more about this feature, and found that if we have `EnableConfigWatch` a config in the `ConfigProviderSettings`, then send a "nil" watcher that does disables "watching" completely.

What do you think?

File: service/collector.go
Comment (swiatekm-sumo) [reply service/collector.go]: Is there any situation where this is actually going to be spam? To begin with, it should be very rare for a config source to change very frequently. My thinking was that this would rarely produce an unwanted message for some users, and catch misconfigurations for others. If you think that's not a good tradeoff, I can remove it, it's not that important.

File: service/collector.go
Comment (swiatekm-sumo) [reply service/collector.go:175]: I thought about it, and ended up putting this in the collector loop for two reasons:
* It's more visible and explicit this way. If it was part of `ConfigProvider` logic, then we'd have a case in the collector loop that implicitly wouldn't do anything depending on settings.
* It's possible for distributions to provide their own `ConfigProvider` implementations, and in that case we'd be forcing them to implement this logic on their own.

That said, I don't think it really matters *that* much where it is.

Comment (bogdandrutu) [reply service/collector.go:175]: I would go this way for the moment... Feels more natural to be there... Happy to hear other opinions.

**Comment by swiatekm-sumo** [comment]
@bogdandrutu are there any other concerns I should address before we can get this merged?

File: service/collector.go
Comment (jpkrohling) [reply service/collector.go]: INFO level would be more appropriate. On Kubernetes, those events can be quite frequent, as there are 3 or 4 levels of symlink between the mount path and the actual file, and any changes to that would trigger an event.

File: service/collector.go
Comment (jpkrohling) [reply service/collector.go:175]: I'm with @bogdandrutu here. Disabling (or not enabling) the config watch should not have the file watched at all. I understand the sentiment, but I also prefer to not have this code executed at all unless the config watch is enabled.

File: service/flags.go
Comment (bogdandrutu) [reply service/flags.go]: `--watch-config` is fine

File: service/collector.go
Comment (bogdandrutu) [reply service/collector.go:175]: @swiatekm-sumo we are blocked on this :)

Comment (swiatekm-sumo) [reply service/collector.go:175]: I had misunderstood your original comment @bogdandrutu, and thought you wanted to leave this as is. Sorry about that, I'm fine moving this check to a lower level.

If we want to do no additional work at all if this is disabled, then we'd need to pass this value down the the Resolver and indicate to the provider we don't want notifications by passing a nil watcher function here: https://github.com/open-telemetry/opentelemetry-collector/blob/ae077bec51f81e838bacbb522f5711636f7a83e0/confmap/resolver.go#L312. Does that sound reasonable?

File: service/collector.go
Comment (swiatekm-sumo) [reply service/collector.go]: After moving the change to Resolver, this doesn't exist anymore, as we don't get the notifications to begin with.

**Review by jpkrohling — APPROVED** [review]
LGTM

**Comment by swiatekm-sumo** [comment]
I rebased the changes. `codecov` is complaining about the additional lines in the command handler not being covered by tests, which is correct in that there's no test for it, but I'm not sure if I should add it or leave as is.

**Comment by phrodrigues77** [comment]
Hi! Will this be merged soon?
The flag along with the configuration reloading [5966](https://github.com/open-telemetry/opentelemetry-collector/issues/5966) would be much appreciated!

**Comment by jpkrohling** [comment]
ping @open-telemetry/collector-approvers

**Review by Aneurysm9 — APPROVED** [review]

**Review by dmitryax — APPROVED** [review]
LGTM

**Comment by dmitryax** [comment]
@bogdandrutu are you ok with merging it?

**Comment by dmitryax** [comment]
After giving it a second thought, I don't think this change is necessary. I'd rather lean towards keeping the watch always enabled. But I believe it's important to introduce the ability to watch files as suggested in https://github.com/open-telemetry/opentelemetry-collector/pull/5945 and https://github.com/open-telemetry/opentelemetry-collector/pull/4454.

My reasoning is that I, as a user, may want to watch one file but ignore updates in another. Another concern is that AFAIR, we've been trying to avoid introducing command-line flags and rely on the yaml config as much as possible.

Maybe we can have different config provider prefixes (schema parts) for files that are watched and not? Or have some additional syntax for flags like disable/enable watch in the config provider invocation strings?

@swiatekm-sumo, maybe we bring it for discussion to one of the next SIGs?

**Comment by swiatekm-sumo** [comment]
> After giving it a second thought, I don't think this change is necessary. I'd rather lean towards keeping the watch always enabled. But I believe it's important to introduce the ability to watch files as suggested in #5945 and #4454.
> 
> My reasoning is that I, as a user, may want to watch one file but ignore updates in another. Another concern is that AFAIR, we've been trying to avoid introducing command-line flags and rely on the yaml config as much as possible.
> 
> Maybe we can have different config provider prefixes (schema parts) for files that are watched and not? Or have some additional syntax for flags like disable/enable watch in the config provider invocation strings?
> 
> @swiatekm-sumo, maybe we bring it for discussion to one of the next SIGs?

Sure, let's discuss during the sig. Offline, https://github.com/open-telemetry/opentelemetry-collector/issues/5966 seems to be the best place to talk about this. I actually thought about this as well, and I'm not convinced the automated config reloading is even a good idea (and afaik no other open source monitoring agents do it), and there's a lot of room for different solutions.

Part of the reason the solution in this PR was originally proposed, is that it covers all the providers at once. Even if you introduce new syntax to make this customizable per-file, you still need to cover other providers which may implement notifications in the future.

I'll make a longer and more comprehensive comment about this in the linked issue.

**Pull request closed without merging**
